vm/src/deserializer.h

```diff
@@ -0,0 +1,83 @@
+#ifndef BORGES_DESERIALIZER_H
+#define BORGES_DESERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace Borges {
+
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
+    template <typename T>
+    void deserialize(std::shared_ptr<T>* p);
+
+private:
+    std::size_t size_;
+    const double* data_;
+    std::size_t index_;
+    std::unordered_map<std::size_t, std::shared_ptr<void>> index_to_ptr_map_;
+};
+
+void deserialize(Deserializer& d, bool* p);
+void deserialize(Deserializer& d, double* p);
+void deserialize(Deserializer& d, std::size_t* p);
+
+template <typename T>
+T
+deserialize(Deserializer& d)
+{
+    alignas(T) char b[sizeof(T)];
+    T* p = reinterpret_cast<T*>(b);
+    deserialize(d, p);
+    return *p;
+}
+
+template <typename T>
+void
+Deserializer::deserialize(std::shared_ptr<T>* p)
+{
+    std::size_t i = Borges::deserialize<std::size_t>(*this);
+    auto it = index_to_ptr_map_.find(i);
+    if (it != index_to_ptr_map_.end()) {
+        new (p) std::shared_ptr<T>(it->second);
+        return;
+    }
+    std::shared_ptr<T> ptr(std::make_shared<T>(deserialize<T>(*this)));
+    index_to_ptr_map_.insert(std::make_pair(i, std::shared_ptr<void>(ptr)));
+    new (p) std::shared_ptr<T>(std::move(ptr));
+}
+
+template<typename T>
+void
+deserialize(Deserializer& d, std::unique_ptr<T>* p)
```

Comment:
why don't you pass by (const) reference here ?

Reply:
Because deserialize functions always call methods that need to change the state of the deserializer (most of the time, bumping the index into the buffer), and thus are non-const.

Reply:
no, i mean `std::unique_ptr<T>* p` instead of `const std::unique_ptr<T> & p` - probably because of the new (p) magic seen before.

Reply:
Exactly. Placement new requires a pointer. It can't use a reference. In general, I almost always use references, except when dealing with low level memory apis like this.

Reply:
yup, valid reason, thx.